pkg/flow/flow_components.go

```diff
@@ -0,0 +1,240 @@
+	Label string      // Component label. Not set for singleton components.
+
+	// References and ReferencedBy are the list of IDs in the same module that
+	// this component depends on, or is depended on by, respectively.
+	References, ReferencedBy []string
+
+	Registration component.Registration // Component registration.
+	Health       component.Health       // Current component health.
+
+	Arguments component.Arguments // Current arguments value of the component.
+	Exports   component.Exports   // Current exports value of the component.
+	DebugInfo interface{}         // Current debug info of the component.
+}
+
+// MarshalJSON returns a JSON representation of cd. The format of the
+// representation is not stable and is subject to change.
+func (cd *ComponentDetail) MarshalJSON() ([]byte, error) {
+	type (
+		componentHealthJSON struct {
+			State       string    `json:"state"`
+			Message     string    `json:"message"`
+			UpdatedTime time.Time `json:"updatedTime"`
+		}
+
+		componentDetailJSON struct {
+			Name         string               `json:"name,omitempty"`
+			Type         string               `json:"type,omitempty"`
+			ID           string               `json:"id,omitempty"`
+			Label        string               `json:"label,omitempty"`
+			References   []string             `json:"referencesTo"`
+			ReferencedBy []string             `json:"referencedBy"`
+			Health       *componentHealthJSON `json:"health"`
+			Original     string               `json:"original"`
+			Arguments    json.RawMessage      `json:"arguments,omitempty"`
+			Exports      json.RawMessage      `json:"exports,omitempty"`
+			DebugInfo    json.RawMessage      `json:"debugInfo,omitempty"`
+		}
+	)
+
+	var (
+		references   = cd.References
+		referencedBy = cd.ReferencedBy
+
+		arguments, exports, debugInfo json.RawMessage
+		err                           error
+	)
+
+	if references == nil {
+		references = []string{}
+	}
+	if referencedBy == nil {
+		referencedBy = []string{}
+	}
+
+	arguments, err = riverjson.MarshalBody(cd.Arguments)
+	if err != nil {
+		return nil, err
+	}
+	exports, err = riverjson.MarshalBody(cd.Exports)
+	if err != nil {
+		return nil, err
+	}
+	debugInfo, err = riverjson.MarshalBody(cd.DebugInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(&componentDetailJSON{
+		Name:         cd.Registration.Name,
+		Type:         "block",
+		ID:           cd.ID.LocalID, // TODO(rfratto): support getting component from module.
+		Label:        cd.Label,
+		References:   references,
+		ReferencedBy: referencedBy,
+		Health: &componentHealthJSON{
+			State:       cd.Health.Health.String(),
+			Message:     cd.Health.Message,
+			UpdatedTime: cd.Health.UpdateTime,
+		},
+		Arguments: arguments,
+		Exports:   exports,
+		DebugInfo: debugInfo,
+	})
+}
+
+// ComponentDetailOptions is used by [Flow.ListComponents] and
+// [Flow.GetComponent] to determine how much information
+type ComponentDetailOptions struct {
```

Comment:
Should the `ComponentDetailOptions`  be passed along with the `ComponentDetail` so that state of what was retrieved doesn't get lost if the `ComponentDetail` is passed around? I am unsure if there is a way to tell a `nil` Exports from a didn't retrieve exports.

Reply:
IMO this should be the responsibility of the caller to expose what detail was included if they want to propagate that information downstream to _their_ callers.

I'd like to stat with not including the options and revisit later if it's confusing. WDYT?